### ISSUE TITLE
"import logging" would cause 8+ seconds delay?

### ISSUE THREAD

**Issue by rayluo** [brython-dev/brython#1607]
My first brython project is functionally correct, but each start would take a slow 8 to 10 seconds from page load to my app functioning. Shrinking the app source code during troubleshooting, eventually leads me to this finding. Merely an `import logging` line in a python script would cause 8+ seconds lead time on my machine. Searching around this issues section, I found @petko10 encountered [same issue](https://github.com/brython-dev/brython/issues/1528#issue-734031569):

```python
import logging
# It takes 3-4 seconds
...
```

Petko's computer is probably faster than my old laptop, but the issue is the same: a mere `import` is supposed to be fast. @PierreQuentel would you mind take a look at this?

PS: I understand that, some module might do lots of initialization at module-wide level, thus could be slow. If that is the case, I can try to avoid using `logging` in my own Brython project code, but other 3rd-party package might still use it.

**Comment by PierreQuentel** [comment]
@rayluo 
Am I right if I suppose that you didn't include __`brython_stdlib.js`__ in the page ?

If you didn't, Brython sends an Ajax call for each imported module, and __`logging`__ must be one of those Python packages that import dozens of other modules. Each module is then translated on the fly to Javascript, and the whole process takes seconds.

With __`brython_stdlib.js`__, the Python source code of each module in the stdlib is already loaded (no Ajax call); on the first run, the Javascript translation has to be done, but it is stored in an indexedDB cache, so the translation is not done again on subsequent runs.

This means that the first run takes much less time than without __`brython_stdlib.js`__, and that subsequent runs load the module almost immediately.

**Comment by rayluo** [comment]
Thanks, Pierre, for the detailed explanation. I am aware of those effects, as you also did a good job in documenting them in [here](https://brython.info/static_doc/en/import.html).

But my current case includes the `brython_stdlib.js`, and I can even confirm that by the inspector of browser: `brython_stdlib.js` was loaded, and then there was no extra ajax calls.

I suspect you can also try to reproduce this by creating a default project with `brython.js` & `brython_stdlib.js` etc., but the project code contains only this, and then you can check your console log:

```python
print("start import")
import logging
print("end import")
```

Unless your computer is super fast, the delay would be tangible. Or you can time it and compare it with importing some other modules.

**Comment by PierreQuentel** [comment]
Ah, this is more complex than I thought...

The only possibility is that for some reason the indexedDB cache is not used.

Could you set the debug mode to 2 ? In this case the first message in the console should be
```
using indexedDB for stdlib modules cache
```

On the first run, you should see a number of messages such as `precompile <module>` and `run Python code from VFS <module>`, then messages such as `<module> stored in db`.

On next runs, you should see messages `load <module> from precompiled`

Is it the case ?

**Comment by rayluo** [comment]
Since I've been working on my such first project for a while, now my run would see output like these:

```
11:17:29.334 using indexedDB for stdlib modules cache brython.min.js:1:169564
11:17:29.616 load from cache __future__ brython.min.js:1:170029
11:17:29.617 load from cache _codecs
...
11:17:30.172 load from cache zipimport brython.min.js:1:170029
11:17:30.174 load from cache zlib brython.min.js:1:170029
11:17:30.176 done

NOTE: There are 3 long brython code snippets here. Let me know you would also need them.

11:17:30.717 load json from precompiled brython.min.js:1:351230
11:17:30.745 load collections from precompiled brython.min.js:1:351230
11:17:30.833 load _collections_abc from precompiled brython.min.js:1:351230

NOTE: WAS THE FOLLOWING LINE A PROBLEM?
11:17:30.922 unreachable code after return statement  brython.min.js:1543:16

11:17:30.970 load abc from precompiled brython.min.js:1:351230
11:17:30.986 load _py_abc from precompiled
...
11:17:32.179 load _thread from precompiled brython.min.js:1:351230
11:17:32.199 load _collections from precompiled
```

The entire loading time is around 3 seconds. This is after I removed the `import logging`. I will add it back and then re-run and update this message.

**Comment by PierreQuentel** [comment]
> NOTE: WAS THE FOLLOWING LINE A PROBLEM?
> 11:17:30.922 unreachable code after return statement  brython.min.js:1543:16

No, it's just a warning message that the Javascript parser prints. It happens because of such situations in a few Python scripts from the stdlib. No impact on performance.

**Comment by rayluo** [comment]
After adding:

```python
print("before import logging")
import logging
print("after import logging")
```

9 seconds added to the loading time. But, the console output seems reasonable. Does that mean there would be not much we can improve here?

```
...
12:29:54.702 logging imports atexit,collections.abc,io,os,pickle,re,string,sys,threading,time,traceback,warnings,weakref brython.min.js:1:168764
12:29:54.708 before import logging
brython.min.js:1:207780
12:29:54.716 load logging from precompiled brython.min.js:1:351230
12:29:54.907 load sys from precompiled brython.min.js:1:351230
12:29:55.032 load os from precompiled brython.min.js:1:351230
12:29:55.044 load abc from precompiled brython.min.js:1:351230
12:29:55.059 load _py_abc from precompiled brython.min.js:1:351230
12:29:55.068 load _weakrefset from precompiled brython.min.js:1:351230
12:29:55.087 load _weakref from precompiled brython.min.js:1:351230
12:29:55.101 load types from precompiled brython.min.js:1:351230
12:29:55.264 load posixpath from precompiled brython.min.js:1:351230
12:29:55.283 load stat from precompiled brython.min.js:1:351230
12:29:55.302 load genericpath from precompiled brython.min.js:1:351230
12:29:55.365 load time from precompiled brython.min.js:1:351230
12:29:55.453 load io from precompiled brython.min.js:1:351230
12:29:55.458 load _io from precompiled brython.min.js:1:351230
12:29:55.465 load codecs from precompiled brython.min.js:1:351230
12:29:55.562 load _codecs from precompiled brython.min.js:1:351230
12:29:55.655 load errno from precompiled brython.min.js:1:351230
12:29:55.895 load _thread from precompiled brython.min.js:1:351230
12:29:56.083 load re from precompiled brython.min.js:1:351230
12:29:56.099 load enum from precompiled brython.min.js:1:351230
12:29:56.602 load sre_compile from precompiled brython.min.js:1:351230
12:29:56.637 load _sre from precompiled brython.min.js:1:351230
12:29:56.721 load operator from precompiled brython.min.js:1:351230
12:29:56.772 load _operator from precompiled brython.min.js:1:351230
12:29:56.776 load sre_constants from precompiled brython.min.js:1:351230
12:29:57.136 load sre_parse from precompiled brython.min.js:1:351230
12:29:57.321 load functools from precompiled brython.min.js:1:351230
12:29:57.373 load collections from precompiled brython.min.js:1:351230
12:29:57.461 load _collections_abc from precompiled brython.min.js:1:351230
12:29:57.536 unreachable code after return statement
brython.min.js:1543:16
12:29:58.285 load heapq from precompiled brython.min.js:1:351230
12:29:58.311 load itertools from precompiled brython.min.js:1:351230
12:29:58.402 load keyword from precompiled brython.min.js:1:351230
12:29:58.405 load reprlib from precompiled brython.min.js:1:351230
12:29:58.425 load _collections from precompiled brython.min.js:1:351230
12:29:59.004 load _functools from precompiled brython.min.js:1:351230
12:29:59.408 load copyreg from precompiled brython.min.js:1:351230
12:29:59.428 load traceback from precompiled brython.min.js:1:351230
12:29:59.451 load linecache from precompiled brython.min.js:1:351230
12:29:59.458 load tokenize from precompiled brython.min.js:1:351230
12:29:59.492 load token from precompiled brython.min.js:1:351230
12:30:00.458 load warnings from precompiled brython.min.js:1:351230
12:30:00.495 load weakref from precompiled brython.min.js:1:351230
12:30:00.729 load collections.abc from precompiled brython.min.js:1:351230
12:30:00.733 load string from precompiled brython.min.js:1:351230
12:30:01.114 load threading from precompiled brython.min.js:1:351230
12:30:03.171 load atexit from precompiled brython.min.js:1:351230
12:30:03.203 after import logging
...
```

**Comment by PierreQuentel** [comment]
After the first run, the same code runs in about 2 sec. in Firefox, 2.5 in Chrome, 3 in Edge. Much faster than your result, although my PC is not particularly fast (processor Intel Core 3 GHz, 8 GB RAM). Apart from the performance of your computer, I don't know what can cause such a difference.

**Comment by rayluo** [comment]
Fair enough! I tested it with different computer, and indeed the result varies per computer.

PS: The benefit of using a less-than-awesome computer is that I'd notice potential performance weak point before my end users, so that I can optimize it beforehand. Now, by removing `logging` from my project, I can save "2 sec. in Firefox, 2.5 in Chrome, 3 in Edge" for a typical end user. :-P

Closing this issue.
